src/KOKKOS/kokkos_type.h

```diff
@@ -1076,15 +1076,29 @@ struct params_lj_coul {
 
 // Pair SNAP
 
+#define SNAP_KOKKOS_REAL double
+#define SNAP_KOKKOS_HOST_VECLEN 1
+
+#ifdef LMP_KOKKOS_GPU
+#define SNAP_KOKKOS_DEVICE_VECLEN 32
+#else
+#define SNAP_KOKKOS_DEVICE_VECLEN 1
+#endif
+
+
+// intentional: SNAreal/complex gets reused beyond SNAP
 typedef double SNAreal;
 
 //typedef struct { SNAreal re, im; } SNAcomplex;
-template <typename real>
-struct alignas(2*sizeof(real)) SNAComplex
+template <typename real_>
+struct alignas(2*sizeof(real_)) SNAComplex
 {
+  using real = real_;
```

Comment:
Not sure why this is needed, i.e. can we just use the template parameter?

Reply:
You can use it inside the definition of the `SNAComplex` class, but not outside the definition. Sort of a contrived example, but you can do something like:

```
typedef SNAComplex<double> my_complex;
[...]
using real = typename my_complex::real; // double
```
But not
```
typedef SNAComplex<double> my_complex;
[...]
using real = typename my_complex::real_; // iirc throws a compile time error
```

All this said---in the changes I've made it isn't strictly necessary, that is, I could have just used `real` everywhere and it would have been fine, but that's only because everything uses a uniform precision---`PairSNAPKokkos`, `SNAKokkos`, all reductions within it---as opposed to mixed precision. This is more of a forward-thinking change.

Reply:
@weinbe2 @stanmoore1 please note that using "real" as a type or variable name can lead to conflicts with some libraries and legacy codes. If I remember correctly, we used to have a definition of "real" in some header of some LAMMPS package in the past and that created some nasty and hard to debug compilation issues when certain combination of packages were included. This is one of the reasons why I have been consistently pushing for keeping our style headers with the minimum amount of include files and PIMPLify cases where this cannot easily be done.

Even if there is no conflict the use of a common name like "real" can be confusing, especially for people with a significant exposure to Fortran. ;-) Thus making the name more specific, e.g. "real_kokkos" or "real_kk_t" or more descriptive of the use case could help with future maintenance.

Reply:
I just switched `real` to `real_type`, is that acceptable?

Reply:
I don't recall any library/software using "real_type", so yes, that'll do. Ultimately, the real mistake is in those libraries or packages that define "real" in their (global) headers instead of using something more unique. But since those kind of conflicts as so nasty to track down, it is a good idea to be proactive.
In LAMMPS also there is a hierarchy of how much of a problem this can be. Local definitions inside a function in a .cpp file are the least concern (if at all) while a (pair) style header file, which is included together with many other such files at the same time, is the biggest issue. So a package specific header file that is only used by .cpp files in the same package is a lesser risk anyway.